executor/grant.go

```diff
@@ -72,8 +72,15 @@ func (e *GrantExec) Next(ctx context.Context, req *chunk.Chunk) error {
 		dbName = e.ctx.GetSessionVars().CurrentDB
 	}
 
-	// Make sure the table exist.
+	// For table , check whether table exists and privilege is valid
 	if e.Level.Level == ast.GrantLevelTable {
+		// Return if privilege is invalid, to fail before not existing table, see issue #29302
+		for _, p := range e.Privs {
+			// Exclude column level
+			if len(p.Cols) == 0 && !mysql.AllowedTablePrivs.Has(p.Priv) {
```

Comment:
Why excluding column level? If it is not allowed on table level, how could it be allowed on column level?
Please add some more tests for this as well, if you really want to differentiate between column level and table level.

MySQL returns:
```sql
mysql> grant create temporary tables (column_name) on NotExistsD29302.NotExistsT29302 to u29302;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(column_name) on NotExistsD29302.NotExistsT29302 to u29302' at line 1
```
and depending on the check for `len(p.Cols) == 0` it will return Non-existing table or ErrIllegalGrantForTable which is not correct either. I suggest you create a separate bug for the column grants instead.

Reply:
It will return `Incorrect usage of %s and %s`. Actually I would like to exclude column level to execute column check here https://github.com/pingcap/tidb/blob/ec7a638e33fdde89d07f50b7ad587bc00d1d94f0/executor/grant.go#L671-L674

Reply:
Update: I moved the column privilege validation to the same place, before finding table, so it will return `ErrWrongUsage` for this sql stmt. I think this comment can be resolved.